internal/backend/backend.go

```diff
@@ -131,3 +138,90 @@ type resource struct {
 	ScannedAt        metav1.Time   `json:"scanned_at"`
 	DeletedAt        *metav1.Time  `json:"deleted_at,omitempty"`
 }
+
+type upsertFailures struct {
+	failures map[types.UID]*upsertFailure
+	oldest   *upsertFailure
+
+	failureHistogram *prometheus.HistogramVec
+	oldestFailure    *prometheus.GaugeVec
+}
+
+func newUpsertFailures() *upsertFailures {
+	u := &upsertFailures{
+		failures: make(map[types.UID]*upsertFailure),
+		oldest:   nil,
+
+		failureHistogram: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name: "backend_retries",
+				Help: "Number of retries until resources were upserted successfully, partinioned by their last failure code",
+			},
+			[]string{"code"},
+		),
+		oldestFailure: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "backend_oldest_failure",
+			Help: "A timestamp of when the oldest resource has been tried the first time",
+		}, []string{"code"}),
+	}
+	metrics.Registry.MustRegister(u)
+	return u
+}
+
+func (u upsertFailures) recordFailure(code int, resource client.Object) {
```

Comment:
Does anything actually call this or recordSuccess?

Reply:
not yet, it should be called then from the normal backend functions. But I think it's pretty obvious how they're supposed to be used at least :)